### PR TITLE
Tezos: Add voting support

### PR DESCRIPTION
**Pull request by adonagy**
Added support for signing tezos voting operations - proposal and ballot. With these operations we allow tezos user to participate in the protocol amendment process with a Trezor T device.

(With the requested changes from https://github.com/trezor/trezor-core/pull/518 and https://github.com/trezor/trezor-common/pull/300)

### CODE REVIEW

File: core/src/apps/tezos/helpers.py [hunk 1]
```diff
@@ -43,12 +46,22 @@ def validate_full_path(path: list) -> bool:
...
+    length = len(path)
+    if length < 3 or length > 4:
         return False
     if path[0] != 44 | HARDENED:
         return False
     if path[1] != 1729 | HARDENED:
         return False
     if path[2] < HARDENED or path[2] > 1000000 | HARDENED:
         return False
+    if length > 3 and (path[3] < HARDENED or path[3] > 1000000 | HARDENED):
```
Comment (adonagy) [inline core/src/apps/tezos/helpers.py]: Including the 4th component to the derivation path in this PR

File: core/src/apps/tezos/sign_tx.py [hunk 2]
```diff
@@ -171,3 +204,28 @@ def _encode_zarith(w: bytearray, num):
...
         write_uint8(w, 128 | byte)
+
+
+def _encode_proposal(w: bytearray, proposal):
+    proposal_tag = 5
+
+    write_uint8(w, proposal_tag)
+    write_bytes(w, proposal.source)
+    write_uint32_be(w, proposal.period)
+    write_uint32_be(w, _get_proposals_bytecount(proposal.proposals))
```
Comment (matejcik) [inline core/src/apps/tezos/sign_tx.py]: having a function for this seems needless? just inline the `len(proposals) * PROPOSAL_LENGTH` here

File: core/src/apps/tezos/helpers.py (hunk 1 above)
Comment (matejcik) [reply core/src/apps/tezos/helpers.py]: still not too sure about this - you're allowing two distinct components a wide range of values, and also it's a fourth completely separate scheme :(

what paths are used on Ledger?

**Comment by matejcik** [comment]
other than the comments LGTM.

noting that this is contingent on fixing the paginator bug
handing over to @jpochyla for that, and also for final review

File: python/trezorlib/tests/device_tests/test_msg_tezos_sign_tx.py [hunk 3]
```diff
@@ -194,3 +195,140 @@ def test_tezos_sign_tx_delegation(self):
...
+            TEZOS_PATH_10,
+            dict_to_proto(
+                messages.TezosSignTx,
+                {
+                    "branch": "3a8f60c4cd394cee5b50136c7fc8cb157e8aaa476a9e5c68709be6fc1cdb5395",
+                    "ballot": {
+                        "source": "0002298c03ed7d454a101eb7022bc95f7e5f41ac78",
+                        "period": 2,
+                        "proposal": "def7ed9c84af23ab37ebb60dd83cd103d1272ad6c63d4c05931567e65ed027e3",
+                        "ballot": 0
```
[the remark this answers is not in the document]
Comment (matejcik) [reply python/trezorlib/tests/device_tests/test_msg_tezos_sign_tx.py]: no, actually this is correct, this is the input to `dict_to_proto`

File: core/src/apps/tezos/helpers.py (hunk 1 above)
Comment (adonagy) [reply core/src/apps/tezos/helpers.py]: The default path on ledger is m/44'/1729'/0'/0'. See https://github.com/obsidiansystems/ledger-app-tezos/blob/master/README.md#importing-the-key-from-the-ledger-nano-s

**Comment by adonagy** [comment]
I've made the requested changes. Let me know if there is anything else :)

File: core/src/apps/tezos/helpers.py (hunk 1 above)
Comment (matejcik) [reply core/src/apps/tezos/helpers.py]: would it be a problem to fix the 4th path component at `0'`?

Comment (adonagy) [reply core/src/apps/tezos/helpers.py]: The idea is, that we want to keep the option to use the paths up to now. e.g:

m/44'/1729'/0'
m/44'/1729'/1'
m/44'/1729'/2' 
etc.

and also allow ledger user to migrate to trezor. for that we need paths, e.g.:

m/44'/1729'/0'/0'
m/44'/1729'/0'/1'
m/44'/1729'/0'/2'
etc.

So fixing the 4th component would be a problem.

Comment (matejcik) [reply core/src/apps/tezos/helpers.py]: gah, why does Ledger need to do a completely arbitrary thing yet again :(

in that case let's enforce that
a) either `length == 3` and then `0' <= path[2] <= 1M'`
b) or `length == 4` and then `path[2] == 0'` and `0' <= path[3] <= 1M'`

Comment (adonagy) [reply core/src/apps/tezos/helpers.py]: Yeah, that should do the trick :)

File: core/src/apps/tezos/helpers.py [hunk 4]
```diff
@@ -43,12 +46,23 @@ def validate_full_path(path: list) -> bool:
     Validates derivation path to equal 44'/1729'/a',
     where `a` is an account index from 0 to 1 000 000.
     """
-    if len(path) != 3:
+    length = len(path)
+    print(path)
```
Comment (matejcik) [inline core/src/apps/tezos/helpers.py]: stray print

File: core/src/apps/tezos/helpers.py [hunk 5]
```diff
@@ -43,12 +46,23 @@ def validate_full_path(path: list) -> bool:
...
+    if length < 3 or length > 4:
         return False
     if path[0] != 44 | HARDENED:
         return False
     if path[1] != 1729 | HARDENED:
         return False
-    if path[2] < HARDENED or path[2] > 1000000 | HARDENED:
+    if length == 3 and (path[2] < HARDENED or path[2] > 1000000 | HARDENED):
+        return False
+    if length == 4 and (path[2] != 0 | HARDENED or path[3] < HARDENED or path[3] > 1000000 | HARDENED):
```
Comment (matejcik) [inline core/src/apps/tezos/helpers.py]: this line is too long
maybe just fork based on length?
```python
if length == 3:
    if path[2]...:
        return False
elif length == 4:
    if path[2] != 0 | HARDENED:
        return False
    if path[3]...:
        return False
```

Comment (adonagy) [reply core/src/apps/tezos/helpers.py]: > our CI does not do this right now, so please make sure to run black over all your python files

Sure, will do.

**Comment by adonagy** [comment]
> our CI does not do this right now, so please make sure to run black over all your python files

Sure, will do

File: core/src/apps/tezos/helpers.py (hunk 5 above)
Comment (adonagy) [reply core/src/apps/tezos/helpers.py]: Yeah, makes it more readable. Will fix it ASAP

**Comment by adonagy** [comment]
Run black on all the modified python files. 1 file was reformatted (`test_msg_tezos_sign_tx.py`)

**Pull request merged**